presto-main/src/main/java/io/prestosql/operator/HashSemiJoinOperator.java

```diff
@@ -78,118 +80,151 @@ public OperatorFactory duplicate()
-    public void addInput(Page page)
+    public void close()
+            throws Exception
     {
-        requireNonNull(page, "page is null");
-        checkState(!finishing, "Operator is finishing");
-        checkState(channelSet != null, "Set has not been built yet");
-        checkState(outputPage == null, "Operator still has pending output");
-
-        // create the block builder for the new boolean column
-        // we know the exact size required for the block
-        BlockBuilder blockBuilder = BOOLEAN.createFixedSizeBlockBuilder(page.getPositionCount());
-
-        Page probeJoinPage = new Page(page.getBlock(probeJoinChannel));
-        Optional<Block> hashBlock = probeHashChannel.map(page::getBlock);
-
-        // update hashing strategy to use probe cursor
-        for (int position = 0; position < page.getPositionCount(); position++) {
-            if (probeJoinPage.getBlock(0).isNull(position)) {
-                if (channelSet.isEmpty()) {
-                    BOOLEAN.writeBoolean(blockBuilder, false);
-                }
-                else {
-                    blockBuilder.appendNull();
-                }
+    }
+
+    private class SemiJoinPages
+            implements WorkProcessor.Transformation<Page, Page>
+    {
+        private final int probeJoinChannel;
+        private final ListenableFuture<ChannelSet> channelSetFuture;
+        private final Optional<Integer> probeHashChannel;
+
+        private ChannelSet channelSet;
+
+        public SemiJoinPages(SetSupplier channelSetFuture, int probeJoinChannel, Optional<Integer> probeHashChannel)
+        {
+            requireNonNull(channelSetFuture, "hashProvider is null");
+            checkArgument(probeJoinChannel >= 0, "probeJoinChannel is negative");
+
+            this.channelSetFuture = channelSetFuture.getChannelSet();
+            this.probeJoinChannel = probeJoinChannel;
+            this.probeHashChannel = probeHashChannel;
+        }
+
+        @Override
+        public WorkProcessor.TransformationState<Page> process(Page inputPage)
+        {
+            if (!this.channelSetFuture.isDone()) {
+                return WorkProcessor.TransformationState.blocked(this.channelSetFuture);
             }
-            else {
-                boolean contains;
-                if (hashBlock.isPresent()) {
-                    long rawHash = BIGINT.getLong(hashBlock.get(), position);
-                    contains = channelSet.contains(position, probeJoinPage, rawHash);
-                }
-                else {
-                    contains = channelSet.contains(position, probeJoinPage);
-                }
-                if (!contains && channelSet.containsNull()) {
-                    blockBuilder.appendNull();
-                }
-                else {
-                    BOOLEAN.writeBoolean(blockBuilder, contains);
-                }
+
+            if (channelSet == null) {
+                channelSet = tryGetFutureValue(channelSetFuture).orElse(null);
```

Comment:
future must be done at this point, so `tryGetFutureValue` is not needed.
You can use `io.airlift.concurrent.MoreFutures#checkSuccess` instead